### PR TITLE
Initial implementation of MSC3981: recursive relations API

### LINKED ISSUE #15377: Implement MSC3981 in Synapse
The proposal https://github.com/matrix-org/matrix-spec-proposals/pull/3981 requires an implementation in Synapse.

This is already happening in https://github.com/matrix-org/synapse/pull/15315.

### PR DESCRIPTION
**Pull request by clokep**
Implements [MSC3981](https://github.com/matrix-org/matrix-spec-proposals/pull/3981) which adds a `recurse` keyword argument to the `/relations` API to include children of the requested event recursively.

Fixes #15377

### CODE REVIEW

**Comment by clokep** [comment]
(Note this is waiting for potential changes in the MSC.)

File: synapse/storage/databases/main/relations.py [hunk 1]
```diff
@@ -200,7 +202,7 @@ async def get_relations_for_event(
         # Ensure bad limits aren't being passed in.
         assert limit >= 0
 
-        where_clause = ["relates_to_id = ?", "room_id = ?"]
+        where_clause = ["room_id = ?"]
         where_args: List[Union[str, int]] = [event.event_id, room_id]
```
Comment (erikjohnston) [inline synapse/storage/databases/main/relations.py:206]: It's quite confusing that the `where_clause` and `where_args` no longer line up

File: synapse/storage/databases/main/relations.py [hunk 2]
```diff
@@ -229,18 +231,42 @@ async def get_relations_for_event(
...
-        """ % (
-            " AND ".join(where_clause),
-            order,
-            order,
-        )
+        # TODO This needs to be a recursive query that maybe still matches some of the times above.
+        if recurse:
+            sql = """
+                WITH RECURSIVE related_events AS (
+                    SELECT event_id, relation_type, relates_to_id, 0 depth
```
Comment (erikjohnston) [inline synapse/storage/databases/main/relations.py:238]:
```suggestion
                    SELECT event_id, relation_type, relates_to_id, 0 AS depth
```

for my sanity

Comment (clokep) [reply synapse/storage/databases/main/relations.py:238]: Should I make this change in the other similar queries we have too? 😄

Comment (erikjohnston) [reply synapse/storage/databases/main/relations.py:238]: Ah. I'd certainly welcome it :smile:

File: synapse/storage/databases/main/relations.py [hunk 3]
```diff
@@ -229,18 +231,42 @@ async def get_relations_for_event(
...
-            INNER JOIN events USING (event_id)
-            WHERE %s
-            ORDER BY topological_ordering %s, stream_ordering %s
-            LIMIT ?
-        """ % (
-            " AND ".join(where_clause),
-            order,
-            order,
-        )
+        # TODO This needs to be a recursive query that maybe still matches some of the times above.
```
Comment (erikjohnston) [inline synapse/storage/databases/main/relations.py:234]: I'm assuming we need to fix this before we can merge?

Comment (clokep) [reply synapse/storage/databases/main/relations.py:234]: Errr, I have no idea what this comment is saying. 🤨

File: synapse/storage/databases/main/relations.py [hunk 4]
```diff
@@ -229,23 +231,52 @@ async def get_relations_for_event(
...
+                    FROM event_relations e
+                    INNER JOIN related_events r ON r.event_id = e.relates_to_id
+                    WHERE depth <= 3
+                )
+                SELECT event_id, relation_type, sender, topological_ordering, stream_ordering
+                FROM related_events
+                INNER JOIN events USING (event_id)
+                WHERE %s
+                ORDER BY topological_ordering %s, stream_ordering %s
+                LIMIT ?;
```
Comment (clokep) [inline synapse/storage/databases/main/relations.py:256]: The `SELECT` portion of these queries are the same -- would be better / clearer to structure this as a preamble & table name to query and only have one copy of the shared bit?

Comment (erikjohnston) [reply synapse/storage/databases/main/relations.py:256]: I think this is probably fine for now

**Review by erikjohnston — APPROVED** [review]
Sorry for missing this last week!

**Pull request merged**